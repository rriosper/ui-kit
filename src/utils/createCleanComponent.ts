import { omit } from 'ramda'
import { createElement } from 'react'

type UncleanComponentProps = {
  children: React.ReactNode;
  [key: string]: any;
};

// Function to create cleaned component witouth any invalid html property.
const createCleanComponent = (
  Component: React.FunctionComponent | string,
  propsToRemove: string[]
) => ({ children, ...props }: UncleanComponentProps) => {
  const filteredProps = omit(propsToRemove, props)
  return createElement(Component, filteredProps, children)
}

export default createCleanComponent
