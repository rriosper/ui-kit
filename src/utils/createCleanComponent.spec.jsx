import { render } from "@testing-library/react";
import styled from "@xstyled/styled-components";
import { backgroundColor } from "@xstyled/system";
import React from "react";
import createCleanComponent from "./createCleanComponent";

const CleanComponent = createCleanComponent("div", [
  "testInvalidProp",
  "invalidProp",
  "backgroundColor",
]);

const Component = styled(CleanComponent)`
  ${backgroundColor}
  width: 100px;
  height: 100px;
  background: red;
`;

describe("createCleanComponent", () => {
  test("Without children", () => {
    const { asFragment, getByTestId } = render(
      <Component
        data-testid="component"
        disabled
        testInvalidProp="testInvalidProp"
        invalidProp="invalidProp"
        backgroundColor="red"
      />
    );

    expect(getByTestId("component")).not.toHaveAttribute("testInvalidProp");
    expect(getByTestId("component")).not.toHaveAttribute("invalidProp");
    expect(getByTestId("component")).not.toHaveAttribute("backgroundColor");
    expect(getByTestId("component")).toHaveAttribute("disabled");

    expect(asFragment()).toMatchSnapshot();
  });

  test("With children", () => {
    const { asFragment, getByTestId, getByText } = render(
      <Component
        data-testid="component"
        disabled
        testInvalidProp="testInvalidProp"
        invalidProp="invalidProp"
        backgroundColor="red"
      >
        CHILDREN!
      </Component>
    );

    expect(getByTestId("component")).not.toHaveAttribute("testInvalidProp");
    expect(getByTestId("component")).not.toHaveAttribute("invalidProp");
    expect(getByTestId("component")).not.toHaveAttribute("backgroundColor");
    expect(getByTestId("component")).toHaveAttribute("disabled");
    expect(getByText("CHILDREN!")).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
