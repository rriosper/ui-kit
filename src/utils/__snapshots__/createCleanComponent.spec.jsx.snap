// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCleanComponent With children 1`] = `
<DocumentFragment>
  .c0 {
  background-color: red;
  width: 100px;
  height: 100px;
  background: red;
}

<div
    class="c0"
    data-testid="component"
    disabled=""
  >
    CHILDREN!
  </div>
</DocumentFragment>
`;

exports[`createCleanComponent Without children 1`] = `
<DocumentFragment>
  .c0 {
  background-color: red;
  width: 100px;
  height: 100px;
  background: red;
}

<div
    class="c0"
    data-testid="component"
    disabled=""
  />
</DocumentFragment>
`;
